File: src/622_Design_Circular_Queue.py
# 622_Design_Circular_Queue
# https://leetcode.cn/problems/design-circular-queue/

class Node:
    def __init__(self, v: int):
        self.pre = None
        self.nxt = None
        self.val = v

class MyCircularQueue:

    def __init__(self, k: int):
        self.cnt = 0
        self.capacity = k
        self.dummy = Node(0)  # Initialize with a dummy node
        self.dummy.nxt = self.dummy
        self.dummy.pre = self.dummy

    def enQueue(self, value: int) -> bool:
        if self.cnt == self.capacity:
            return False
        
        node = Node(value)
        last_node = self.dummy.pre
        last_node.nxt = node
        node.pre = last_node
        node.nxt = self.dummy
        self.dummy.pre = node
        
        self.cnt += 1
        
        return True

    def deQueue(self) -> bool:
        if self.cnt == 0:
            return False
        
        second_node = self.dummy.nxt.nxt
        self.dummy.nxt = second_node
        second_node.pre = self.dummy
        
        self.cnt -= 1
         
        return True

    def Front(self) -> int:
        if self.cnt == 0:
            return -1
        return self.dummy.nxt.val

    def Rear(self) -> int:
        if self.cnt == 0:
            return -1
        return self.dummy.pre.val

    def isEmpty(self) -> bool:
        return self.cnt == 0

    def isFull(self) -> bool:
        return self.cnt == self.capacity




# Your MyCircularQueue object will be instantiated and called as such:
# obj = MyCircularQueue(k)
# param_1 = obj.enQueue(value)
# param_2 = obj.deQueue()
# param_3 = obj.Front()
# param_4 = obj.Rear()
# param_5 = obj.isEmpty()
# param_6 = obj.isFull()

File: src/622_Design_Circular_Queue.rs
// 622_Design_Circular_Queue
// https://leetcode.cn/problems/design-circular-queue/

// 使用数组模拟循环队列

struct MyCircularQueue {
    queue: Vec<i32>,
    front: usize,
    rear: usize,
    count: usize,
    capacity: usize,
}

impl MyCircularQueue {
    fn new(k: i32) -> Self {
        MyCircularQueue {
            queue: vec![0; k as usize],
            front: 0,
            rear: 0,
            count: 0,
            capacity: k as usize,
        }
    }

    fn en_queue(&mut self, value: i32) -> bool {
        if self.is_full() {
            return false;
        }
        self.queue[self.rear] = value;
        self.rear = (self.rear + 1) % self.capacity;
        self.count += 1;
        true
    }

    fn de_queue(&mut self) -> bool {
        if self.is_empty() {
            return false;
        }
        self.front = (self.front + 1) % self.capacity;
        self.count -= 1;
        true
    }

    fn front(&self) -> i32 {
        if self.is_empty() {
            -1
        } else {
            self.queue[self.front]
        }
    }

    fn rear(&self) -> i32 {
        if self.is_empty() {
            -1
        } else {
            self.queue[(self.rear + self.capacity - 1) % self.capacity]
        }
    }

    fn is_empty(&self) -> bool {
        self.count == 0
    }

    fn is_full(&self) -> bool {
        self.count == self.capacity
    }
}